ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/processor/BlockReplay.java

```diff
@@ -41,13 +44,107 @@ public BlockReplay(
     this.worldStateArchive = worldStateArchive;
   }
 
+  public BlockTrace block(final Hash blockHash, final Action<TransactionTrace> action) {
+    return performActionPerBlockTransaction(
+            blockHash,
+            (body, header, blockchain, mutableWorldState, transactionProcessor) -> {
+              List<TransactionTrace> transactionTraces =
+                  body.getTransactions().stream()
+                      .map(
+                          transaction ->
+                              action.performAction(
+                                  transaction,
+                                  header,
+                                  blockchain,
+                                  mutableWorldState,
+                                  transactionProcessor))
+                      .collect(Collectors.toList());
+              return Optional.of(new BlockTrace(transactionTraces));
+            })
+        .orElse(null);
```

Comment:
I suspect we could avoid duplication through this class by setting up the versions of methods that accept a `Hash` to just lookup the block by hash and if it's present delegate to the version that accepts a `Block` rather than duplicating code.

Reply:
The block is provided as input to `DebugTraceBlock` method so that's why there is a signature with `Block` as an argument,  do we assume that provided block is already present ?